synapse/federation/federation_server.py

```diff
@@ -1327,72 +1327,53 @@ async def check_server_matches_acl(self, server_name: str, room_id: str) -> None
         acl_event = await self._storage_controllers.state.get_current_state_event(
             room_id, EventTypes.ServerACL, ""
         )
-        if not acl_event or server_matches_acl_event(server_name, acl_event):
-            return
-
-        raise AuthError(code=403, msg="Server is banned from room")
+        if acl_event:
+            server_acl_evaluator = await self._get_server_acl_evaluator(
+                acl_event.event_id, acl_event
+            )
+            if not server_acl_evaluator.server_matches_acl_event(server_name):
+                raise AuthError(code=403, msg="Server is banned from room")
 
+    @cached(uncached_args=("acl_event",))
+    def _get_server_acl_evaluator(
+        self, event_id: str, acl_event: EventBase
+    ) -> ServerAclEvaluator:
```

Comment:
I kind of wanted to cache this per-room, but that was proving difficult to do (as this isn't a datastore so you don't get automatic cache invalidation easily). I can look deeper into that though if we think it would be useful.

I also looked at moving this method *to* a datastore, but that means moving the `get_current_state_event` call into it, which depends on the storage controllers, which wrap datastores.

Any thoughts would be appreciated.

Reply:
Also -- I wonder if this will break if the event is redacted? We probably do want to ensure we invalidate this properly.

Reply:
119056d98ab2c41ec5b610b9b08d80a8495b523a implements this, but it somewhat is weird in that the datastore now calls back out to the storage layer to invalidate it.